internal/logging/logger.go

```diff
@@ -73,125 +74,145 @@ var (
+				loopLogger.Info().Msg("setting log level for ALL scopes from environment")
+				l.defaultLogLevelFromEnv = envLevel
+			} else {
+				// if not "all", parse as comma-separated list of scopes
+				scopes := strings.SplitSeq(env, ",")
+				for scope := range scopes {
+					loopLogger.Info().Msgf("setting log level for scope %s from environment", scope)
+					l.scopeLevels[scope] = envLevel
+				}
+			}
 		}
+	}
 
-		if strings.ToLower(env) == "all" {
-			l.defaultLogLevelFromEnv = level
-
-			if finalDefaultLogLevel > level {
-				finalDefaultLogLevel = level
-			}
+	l.defaultLogLevel = defaultLogLevel
+	logger.Info().Msgf("default log level starts at %v", l.defaultLogLevel)
 
-			continue
-		}
+	if l.defaultLogLevel > l.defaultLogLevelFromEnv {
+		logger.Info().Msgf("default log level from env %v", l.defaultLogLevelFromEnv)
+		l.defaultLogLevel = l.defaultLogLevelFromEnv
+	}
 
-		scopes := strings.SplitSeq(strings.ToLower(env), ",")
-		for scope := range scopes {
-			l.scopeLevels[scope] = level
-		}
+	if l.defaultLogLevel > l.defaultLogLevelFromConfig {
+		logger.Info().Msgf("default log level from config %v", l.defaultLogLevelFromConfig)
+		l.defaultLogLevel = l.defaultLogLevelFromConfig
 	}
 
-	l.defaultLogLevel = finalDefaultLogLevel
+	logger.Info().Msgf("default log level %v", l.defaultLogLevel)
 }
 
 func (l *Logger) getScopeLoggerLevel(scope string) zerolog.Level {
-	if l.scopeLevels == nil {
-		l.updateLogLevel()
+	if level, ok := l.scopeLevels[scope]; ok {
+		return level
 	}
 
-	var scopeLevel zerolog.Level
-	if l.defaultLogLevelFromConfig != -2 {
-		scopeLevel = l.defaultLogLevelFromConfig
-	}
-	if l.defaultLogLevelFromEnv != -2 {
-		scopeLevel = l.defaultLogLevelFromEnv
+	// if the scope is not in the map, use the default level from the root logger
+	if l.defaultLogLevelFromConfig != unset {
+		return l.defaultLogLevelFromConfig
 	}
 
-	// if the scope is not in the map, use the default level from the root logger
-	if level, ok := l.scopeLevels[scope]; ok {
-		scopeLevel = level
+	if l.defaultLogLevelFromEnv != unset {
+		return l.defaultLogLevelFromEnv
 	}
 
-	return scopeLevel
+	return l.defaultLogLevel
 }
 
-func (l *Logger) newScopeLogger(scope string) zerolog.Logger {
-	scopeLevel := l.getScopeLoggerLevel(scope)
-	logger := l.l.Level(scopeLevel).With().Str("component", scope).Logger()
+func (l *Logger) getLogger(scope string) *zerolog.Logger {
+	if logger, ok := l.scopeLoggers[scope]; ok && logger != nil {
+		return logger
+	}
 
-	return logger
-}
+	l.loggerMutex.Lock()
+	defer l.loggerMutex.Unlock()
 
-func (l *Logger) getLogger(scope string) *zerolog.Logger {
-	logger, ok := l.scopeLoggers[scope]
-	if !ok || logger == nil {
-		scopeLogger := l.newScopeLogger(scope)
-		l.scopeLoggers[scope] = &scopeLogger
+	// double-check after acquiring the lock
+	if logger, ok := l.scopeLoggers[scope]; ok && logger != nil {
+		return logger
 	}
 
-	return l.scopeLoggers[scope]
+	scopeLevel := l.getScopeLoggerLevel(scope)
```

Comment:
If it's a logging scope we haven't seen... we need to calculate from "defaults" then override from scope-specific env settings.